File: app/src/main/java/nl/viasalix/btroosterlite/TimetableFragment.java
/*  BTRooster Lite: Roosterapp voor Calvijn College
 *  Copyright (C) 2017 Rutger Broekhoff <rutger broekhoff three at gmail dot com>
 *                 and Jochem Broekhoff
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.viasalix.btroosterlite;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.InputType;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import static android.content.Context.CONNECTIVITY_SERVICE;

public class TimetableFragment extends Fragment {
    private OnFragmentInteractionListener mListener;

    // Locaties
    String[] locaties = {
            "Goes Klein Frankrijk",
            "Goes Noordhoeklaan",
            "Goes Stationspark",
            "Krabbendijke Appelstraat",
            "Krabbendijke Kerkpolder",
            "Middelburg",
            "Tholen"
    };

    // Locaties die in de request URL moeten te komen staan
    String[] locatiesURL = {
            "Goes",
            "GoesNoordhoeklaan",
            "GoesStationspark",
            "KrabbendijkeAppelstraat",
            "KrabbendijkeKerkpolder",
            "Middelburg",
            "Tholen",
    };

    // Opslag 'roostercode', location en type
    String code = "";
    String location = "";
    String type = "";

    // Variabelen om beschikbare weken weer te geven
    List<String> availableWeeks = new ArrayList<>();
    List<String> availableWeeksNames = new ArrayList<>();

    // Initialiseer benodigde variabelen
    SharedPreferences sharedPreferences;
    View view;
    WebView webView;

    public TimetableFragment() {
        // Lege constructor is nodig om een fragment te kunnen gebruiken
    }

    public static TimetableFragment newInstance() {
        TimetableFragment fragment = new TimetableFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.appbar_mainactivity_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    private boolean online() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getActivity().getSystemService(CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = null;
        if (connectivityManager != null) {
            networkInfo = connectivityManager.getActiveNetworkInfo();
        }

        return (networkInfo != null && networkInfo.isConnected());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_reload:
                loadTimetable(true);
                return true;
            case R.id.action_settings:
                Intent intent = new Intent(getActivity(), SettingsActivity.class);
                getActivity().startActivity(intent);
                break;
            case R.id.action_opensource:
                Intent ossIntent = new Intent(getActivity(), OssLicensesMenuActivity.class);
                ossIntent.putExtra("title", "Open-source licenties");
                getActivity().startActivity(ossIntent);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_timetable, container, false);
        webView = view.findViewById(R.id.web_view);
        return view;
    }

    @Override
    public void onStart() {
        super.onStart();

        Toolbar toolbar = getActivity().findViewById(R.id.toolbar);
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);

        if (((AppCompatActivity) getActivity()).getSupportActionBar() != null) {
            ((AppCompatActivity) getActivity()).getSupportActionBar().setDisplayShowTitleEnabled(false);
        }

        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setDisplayZoomControls(false);

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());

        if (loadSharedPreferences() != 1)
            loadTimetable(true);
    }

    private int loadSharedPreferences() {
        if (!sharedPreferences.contains("code")) {
            showCodeDialog();
            return 1;
        }

        if (!sharedPreferences.contains("location")) {
            showLocationDialog();
            return 1;
        }

        code = sharedPreferences.getString("code", "12345");

        location = sharedPreferences.getString("location", locaties[0]);
        type = getType(code);

        return 0;
    }

    private String getType(String code) {
        String docentPatternInput = "([A-Za-z]){3}";
        String leerlingPatternInput = "([0-9]){5}";

        Pattern docentPattern = Pattern.compile(docentPatternInput);
        Pattern leerlingPattern = Pattern.compile(leerlingPatternInput);

        if (!docentPattern.matcher(code).matches() && !leerlingPattern.matcher(code).matches()) {
            return "c";
        } else if (docentPattern.matcher(code).matches()) {
            return "t";
        } else if (leerlingPattern.matcher(code).matches()) {
            return "s";
        }

        return "none";
    }

    private void showCodeDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("Code");

        final EditText input = new EditText(getActivity());
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        builder.setView(input);

        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                sharedPreferences.edit().putString("code", input.getText().toString()).apply();
                onStart();
            }
        }).setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                getActivity().finish();
                dialog.cancel();
            }
        });

        builder.show();
    }

    private void showLocationDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("Locatie");

        builder.setItems(locaties, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                sharedPreferences.edit().putString("location", locatiesURL[which]).apply();
                onStart();
            }
        });

        builder.show();
    }

    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(Uri uri);
    }

    private void loadTimetable(boolean getIndexes) {
        webView.getSettings().setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);

        if (online()) {
            webView.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
        } else {
            webView.getSettings().setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
        }

        if (getIndexes) getIndexes();
        else            getTimetable();
    }

    private void getTimetable() {
        int weekChange = sharedPreferences.getInt("t_weekChange", 0);

        String typeString = getType(code);

        Uri.Builder builder = new Uri.Builder();
        builder.scheme("https")
                .authority(MainActivity.AUTHORITY)
                .appendPath("RoosterEmbedServlet")
                .appendQueryParameter("code", code)
                .appendQueryParameter("locatie", location)
                .appendQueryParameter("type", typeString)
                .appendQueryParameter("week", availableWeeks.get(weekChange));
        String url = builder.build().toString();

        webView.loadUrl(url);
    }

    private void getIndexes() {
        if (online()) {
            RequestQueue queue = Volley.newRequestQueue(getActivity());
            Uri.Builder builder = new Uri.Builder();
            builder.scheme("https")
                    .authority(MainActivity.AUTHORITY)
                    .appendPath("RoosterEmbedServlet")
                    .appendQueryParameter("indexOphalen", "1")
                    .appendQueryParameter("locatie", location);
            String url = builder.build().toString();

            StringRequest stringRequest = new StringRequest(Request.Method.GET, url,
                    new Response.Listener<String>() {
                        @Override
                        public void onResponse(String response) {
                            sharedPreferences.edit().putString("t_indexes", response).apply();
                            Log.d("or", response);
                            handleResponse(response);
                        }
                    }, new Response.ErrorListener() {
                @Override
                public void onErrorResponse(VolleyError error) {
                    Log.d("error", error.getMessage());
                }
            });

            queue.add(stringRequest);
        } else {
            String response = sharedPreferences.getString("t_indexes", null);
            handleResponse(response);
            getTimetable();
        }
    }

    private void handleResponse(String response) {
        availableWeeks.clear();
        availableWeeksNames.clear();

        int i = 0;

        String[] responses = response.split("\n");

        for (String responseString : responses) {
            if (responseString.trim().length() > 0) {
                String[] responseStringSplit = responseString.split("\\|", 2);

                availableWeeks.add(i, responseStringSplit[0]);
                availableWeeksNames.add(i, responseStringSplit[1]);
                ++i;
            }
        }

        Spinner weekSpinner = getActivity().findViewById(R.id.week_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_dropdown_item, availableWeeksNames);
        weekSpinner.setAdapter(adapter);

        weekSpinner.setSelection(sharedPreferences.getInt("t_weekChange", 1));

        weekSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int position, long id) {
                sharedPreferences.edit().putInt("t_weekChange", position).apply();

                loadTimetable(false);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {}
        });
    }
}

File: app/src/main/java/nl/viasalix/btroosterlite/fragments/TimetableFragment.kt
/*  BTRooster Lite: Roosterapp voor Calvijn College
 *  Copyright (C) 2017 Rutger Broekhoff <rutger broekhoff three at gmail dot com>
 *                 and Jochem Broekhoff
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.viasalix.btroosterlite.fragments

import android.app.Fragment
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.text.InputType
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Spinner

import com.google.android.gms.oss.licenses.OssLicensesMenuActivity

import java.util.ArrayList
import java.util.Calendar
import java.util.Stack
import java.util.regex.Pattern

import android.content.Context.CONNECTIVITY_SERVICE
import nl.viasalix.btroosterlite.R
import nl.viasalix.btroosterlite.activities.SettingsActivity
import nl.viasalix.btroosterlite.timetable.TimetableIntegration

class TimetableFragment : Fragment() {
    private var weekSpinner: Spinner? = null
    // Opslag 'roostercode', location en type
    private var code: String? = ""
    private var location: String? = ""
    private var type = ""
    // Variabelen om beschikbare weken weer te geven
    private var availableWeeks: MutableList<String> = ArrayList()
    private var availableWeeksNames: MutableList<String> = ArrayList()
    // Initialiseer benodigde variabelen
    private var sharedPreferences: SharedPreferences? = null
    private var currentView: View? = null
    private var webView: WebView? = null
    // private var webViewPL: WebView? = null
    private var pagesToLoad = Stack<String>()
    private var mListener: OnFragmentInteractionListener? = null
    private var ttInteg: TimetableIntegration? = null

    private val currentWeekOfYear: Int
        get() = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.appbar_mainactivity_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_reload -> {
                loadTimetable(true)
                return true
            }
            R.id.action_settings -> {
                val intent = Intent(activity, SettingsActivity::class.java)
                activity.startActivity(intent)
            }
            R.id.action_opensource -> {
                val ossIntent = Intent(activity, OssLicensesMenuActivity::class.java)
                ossIntent.putExtra("title", "Open-source licenties")
                activity.startActivity(ossIntent)
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        currentView = inflater.inflate(R.layout.fragment_timetable, container, false)
        webView = currentView!!.findViewById(R.id.web_view)
        weekSpinner = currentView!!.findViewById(R.id.week_spinner)

        return currentView
    }

    override fun onStart() {
        super.onStart()

        val toolbar = activity.findViewById<Toolbar>(R.id.toolbar)
        (activity as AppCompatActivity).setSupportActionBar(toolbar)

        if ((activity as AppCompatActivity).supportActionBar != null)
            (activity as AppCompatActivity)
                    .supportActionBar!!
                    .setDisplayShowTitleEnabled(false)

        webView!!.settings.builtInZoomControls = true
        webView!!.settings.displayZoomControls = false

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)

        if (sharedPreferences!!.getBoolean("t_preload", false))
            ttInteg!!.downloadAvailableTimetables()

        loadSharedPreferences()

        ttInteg = TimetableIntegration(activity!!, location!!, code!!)

        if (loadSharedPreferences() != 1)
            loadTimetable(true)
    }

    private fun loadSharedPreferences(): Int {
        if (!sharedPreferences!!.contains("code")) {
            showCodeDialog()
            return 1
        }

        if (!sharedPreferences!!.contains("location")) {
            showLocationDialog()
            return 1
        }

        code = sharedPreferences!!.getString("code", "12345")

        location = sharedPreferences!!.getString("location", locaties[0])
        type = getType(code)

        return 0
    }

    private fun getType(code: String?): String {
        val docentPatternInput = "([A-Za-z]){3}"
        val leerlingPatternInput = "([0-9]){5}"

        val docentPattern = Pattern.compile(docentPatternInput)
        val leerlingPattern = Pattern.compile(leerlingPatternInput)

        if (!docentPattern.matcher(code!!).matches() && !leerlingPattern.matcher(code).matches()) {
            return "c"
        } else if (docentPattern.matcher(code).matches()) {
            return "t"
        } else if (leerlingPattern.matcher(code).matches()) {
            return "s"
        }

        return "none"
    }

    private fun showCodeDialog() {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle("Code")

        val layout = LinearLayout(activity)
        val layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT)

        layout.layoutParams = layoutParams
        val input = EditText(activity)
        input.inputType = InputType.TYPE_CLASS_TEXT
        input.layoutParams = layoutParams

        val margin_left_right_bottom = getPixelValue(activity, 24)
        val margin_top = getPixelValue(activity, 20)
        layoutParams.setMargins(margin_left_right_bottom,
                margin_top,
                margin_left_right_bottom,
                margin_left_right_bottom)

        layout.addView(input)

        builder.setView(layout)

        builder.setPositiveButton("OK") { _, _ ->
            sharedPreferences!!.edit().putString("code", input.text.toString()).apply()
            onStart()
        }.setNegativeButton("Annuleren") { dialog, _ ->
                    activity.finish()
                    dialog.cancel()
                }

        builder.show()
    }

    private fun showLocationDialog() {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle("Locatie")
        val items = locaties

        builder.setSingleChoiceItems(items, -1) { dialog, which -> }

        builder.setPositiveButton("OK") { dialog, which ->
            val selectedPosition = (dialog as AlertDialog).listView.checkedItemPosition
            sharedPreferences!!.edit().putString("location", locatiesURL[selectedPosition]).apply()
            Log.d("selected", Integer.toString(selectedPosition))
            onStart()
        }

        builder.setNegativeButton("Annuleren") { dialog, which ->
            activity.finish()
            dialog.cancel()
        }

        builder.show()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    private fun loadTimetable(getIndexes: Boolean) {
        if (getIndexes) {
            getIndexes()
        } else {
            getTimetable(sharedPreferences!!.getString("t_week",
                    currentWeekOfYear.toString()))
        }
    }

    private fun getTimetable(week: String) {
        ttInteg!!.loadTimetable(week.toInt(), webView!!)
    }

    private fun getIndexes() {
        ttInteg!!.getIndexes { it, wasOnline ->
            if (wasOnline) {
                handleResponse(it)
                getTimetable(
                        sharedPreferences!!.getString(
                                "t_week",
                                currentWeekOfYear.toString()))
            } else {
                handleResponse(it)
            }
        }
    }

    private fun handleResponse(response: String?) {
        if (response != null) {
            availableWeeks.clear()
            availableWeeksNames.clear()

            var i = 0
            val responses = response.trim().split("\n")

            for (responseString in responses) {
                if (responseString.isNotEmpty()) {
                    val responseStringSplit = responseString.split("|")

                    availableWeeks.add(i, responseStringSplit[0])
                    availableWeeksNames.add(i, responseStringSplit[1])
                    ++i
                }
            }

            if (activity != null) {
                val adapter = ArrayAdapter(
                        activity,
                        android.R.layout.simple_spinner_dropdown_item,
                        availableWeeksNames)

                weekSpinner!!.adapter = adapter

                weekSpinner!!.setSelection(
                        availableWeeks.indexOf(
                                sharedPreferences!!.getString(
                                        "t_week",
                                        Integer.toString(currentWeekOfYear))))

                weekSpinner!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                    override fun onItemSelected(adapterView: AdapterView<*>, view: View, position: Int, id: Long) {
                        sharedPreferences!!.edit().putString("t_week", availableWeeks[position]).apply()

                        loadTimetable(false)
                    }

                    override fun onNothingSelected(adapterView: AdapterView<*>) {}
                }
            }
        }
    }

    internal interface OnFragmentInteractionListener

    companion object {
        // Locaties
        var locaties = arrayOf("Goes Klein Frankrijk", "Goes Noordhoeklaan", "Goes Stationspark", "Krabbendijke Appelstraat", "Krabbendijke Kerkpolder", "Middelburg", "Tholen")
        // Locaties die in de request URL moeten te komen staan
        var locatiesURL = arrayOf("Goes", "GoesNoordhoeklaan", "GoesStationspark", "KrabbendijkeAppelstraat", "KrabbendijkeKerkpolder", "Middelburg", "Tholen")

        private fun getPixelValue(context: Context, dimenId: Int): Int {
            val resources = context.resources
            return TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_DIP,
                    dimenId.toFloat(),
                    resources.displayMetrics).toInt()
        }
    }
}